airflow/jobs/triggerer_job_runner.py

```diff
@@ -314,95 +116,222 @@ def on_kill(self):
+
+    logger_cache: dict[int, TriggerLoggingFactory] = attrs.field(factory=dict, init=False)
+
+    # A list of triggers that we have told the async process to cancel. We keep them here until we receive the
+    # FinishedTriggers message
+    cancelling_triggers: set[int] = attrs.field(factory=set, init=False)
+
+    # Outbound queue of events
+    events: deque[tuple[int, events.TriggerEvent]] = attrs.field(factory=deque, init=False)
+
+    # Outbound queue of failed triggers
+    # failed_triggers: deque[tuple[int, BaseException]] = attrs.field(factory=deque, init=False)
+    failed_triggers: deque[tuple[int, str]] = attrs.field(factory=deque, init=False)
+
+    def is_alive(self) -> bool:
+        # Set by `_service_subprocess` in the loop
+        return self._exit_code is None
+
+    @classmethod
+    def start(  # type: ignore[override]
+        cls,
+        *,
+        job: Job,
+        logger=None,
+        **kwargs,
+    ):
+        proc = super().start(id=job.id, job=job, target=cls.run_in_process, logger=logger, **kwargs)
+
+        msg = StartTriggerer(requests_fd=proc._requests_fd).model_dump_json().encode() + b"\n"
+        proc.stdin.write(msg)
+        return proc
+
+    def _handle_request(self, msg: ToSyncProcess, log: FilteringBoundLogger) -> None:  # type: ignore[override]
+        if isinstance(msg, TriggerStateChanges):
+            log.debug("State change from async process", state=msg)
+            if msg.events:
+                self.events.extend(msg.events)
+            if msg.failures:
+                self.failed_triggers.extend(msg.failures)
+            for id in msg.finished or tuple():
+                self.running_triggers.discard(id)
+                self.cancelling_triggers.discard(id)
+                # TODO: Close logger? Or is deleting it enough
+                self.logger_cache.pop(id, None)
+            return
+
+        raise ValueError(f"Unknown message type {type(msg)}")
+
+    def run_db_loop(self) -> None:
+        """Run synchronously and handle all database reads/writes."""
+        while not self.stop:
+            if not self.is_alive():
+                log.error("Trigger runner thread has died! Exiting.")
                 break
             with Trace.start_span(span_name="triggerer_job_loop", component="TriggererJobRunner") as span:
-                # Clean out unused triggers
-                if span.is_recording():
-                    span.add_event(name="Trigger.clean_unused")
-                Trigger.clean_unused()
-                # Load/delete triggers
-                if span.is_recording():
-                    span.add_event(name="load_triggers")
+                # Wait for up to 1 second for activity
+                self._service_subprocess(1)
+
                 self.load_triggers()
-                # Handle events
-                if span.is_recording():
-                    span.add_event(name="handle_events")
```

Comment:
This was pointless noise -- handle_events is already decorated with `@add_span` so we can almost nothing by having an event too.